src/compiler/scala/tools/nsc/typechecker/StdAttachments.scala

```diff
@@ -165,4 +165,24 @@ trait StdAttachments {
   def markDynamicRewrite(tree: Tree): Tree = tree.updateAttachment(DynamicRewriteAttachment)
   def unmarkDynamicRewrite(tree: Tree): Tree = tree.removeAttachment[DynamicRewriteAttachment.type]
   def isDynamicRewrite(tree: Tree): Boolean = tree.attachments.get[DynamicRewriteAttachment.type].isDefined
+
+  /**
+   * Marks a tree that has been adapted by typer and sets the original tree that was in place before.
+   * 
+   * Keeping track of the original trees were is an important feature for some compiler plugins (like
+   * Scalameta) and the incremental compiler (Zinc). In both cases, adapting trees loses information
+   * in some sense and do not allow external tools to capture some information stored in user-defined
+   * trees that are optimized away by early phases (mostly, typer).
+   * 
+   * See how the absence of this attachment blocks Zinc: https://github.com/sbt/zinc/issues/227.
+   * Related: https://github.com/scala/scala-dev/issues/340.
+   * 
+   * This attachment is, at the moment, only used to keep track of constant-folded constants. It
+   * has a generic wording in the hope that in the future can be reused in the same context to keep
+   * track of other adapted trees.
+   */
+  case class OriginalAdaptedTreeAttachment(original: Tree)
+  def markOriginal(tree: Tree, original: Tree) = tree.updateAttachment(OriginalAdaptedTreeAttachment(original))
+  def unmarkOriginal(tree: Tree): Tree = tree.removeAttachment[OriginalAdaptedTreeAttachment]
+  def isAdaptedTree(tree: Tree): Boolean = tree.attachments.get[OriginalAdaptedTreeAttachment].isDefined
```

Comment:
I personally prefer not to have those methods, they just make you go through an indirection when reading the code.

Reply:
I'll use the more direct style, rename methods/names and update this PR.

Reply:
Ok, as a reminder:  please do not include the ticket in the commit title/body. It should be in the PR description. (Title shoud be short, referencing the issue in the commit message causes noise on every push.)

Reply:
I'm fine with keeping it, I'll get over my OCD that hates seeing the stale commit references pile up until the git GC kicks in at github :-)